cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -428,6 +454,45 @@ ArrayKernelExec ArithmeticExecFromOp(detail::GetTypeId get_id) {
   }
 }
 
+template <template <typename... Args> class KernelGenerator, typename Op>
```

Comment:
This is rather a lot of code being generated; it might be nice if I could figure out the right template setup to consolidate the temporal types with their respective equivalent integral type.

Reply:
There's GeneratePhysicalInteger, though that doesn't include floating point. Perhaps you could add GeneratePhysicalNumeric?

Reply:
I added this. Note that if we want to have `CommonTimestamp()` I don't think this'll fly since we'd need to scale the timestamp values to the common unit before comparison.

Reply:
No, CommonTimestamp indicates that an implicit cast is necessary (which in the case of timestamps includes the scaling). It's not the responsibility of the kernel to execute that implicit cast

Reply:
Ah! Ok, my mistake. I'll add some test cases for that.

Binary types aren't implemented here, though (and would need some work/restructuring to handle).